### ISSUE TITLE
Streaming upload support

### ISSUE THREAD

**Issue by ghost** [VeliovGroup/Meteor-Files#35]
Would be really nice

**Comment by dr-dimitru** [comment]
Could please explain what do you mean?
Cause video/audio streaming is supported.
And any file upload works like readStream

**Comment by NROL39** [comment]
```javascript 
getStreamSomehow().pipe(streamOperation).pipe(Files)
````
basically, a way to pipe data to the file collection

**Comment by dr-dimitru** [comment]
On the client?

**Comment by NROL39** [comment]
Yeah

**Comment by NROL39** [comment]
Basically, a way to expose that stream, maybe a prepareInsert that returns the writable stream

**Comment by dr-dimitru** [comment]
Cold you please leave few use cases here?

**Comment by NROL39** [comment]
on-the-fly encryption/media encoding is what i'm looking for

**Comment by dr-dimitru** [comment]
@NROL39 This is only place suitable for your requirements: https://github.com/VeliovGroup/Meteor-Files/blob/master/files.coffee#L707 , so we can add additional callback which will has `arrayBuffer` or `Uint8Array` as argument, but it will be just small chunk of whole file, are you sure you will be able to encode file by pieces?

Alternatively I can suggest separate modifier function which will take whole file and then return modified version this ready for upload, but will it be efficient?

**Comment by NROL39** [comment]
Well, loading a 5GB file to memory and operating on it is not feasible on mobile/laptop/pretty-much-any-non-workstation-grade-computer, being able to pipe the data through a modifier function means you only need to hold the current chunk in memory.

**Comment by dr-dimitru** [comment]
@NROL39 this is what exactly I'm talking about. Please, confirm what it is possible to encode file - chunk by chunk, and encoder doesn't need the whole file.

**Comment by NROL39** [comment]
it is, i've been using it for a while (although what im currently doing is streaming the file through the encoder then rebuilding the file from data chunks from it

**Comment by dr-dimitru** [comment]
within this package? right?
If so, could you copy-paste here some code for further testing

**Comment by NROL39** [comment]
using the [filestream](https://www.npmjs.com/package/filestream) module for FileReadStream and FIleWriteStream
and a browserify'd version of node.js' crypto library for encryption
````javascript
key = 'supersecretkey';
iv = 'supersecretiv';
cryptor = crypto.createCipher('aes-128-cbc', key, iv);
fstream = new FileReadStream(file)
fstream.pipe(cryptor).pipe(streamUploadFunction)
````
where file is a [File](https://developer.mozilla.org/en/docs/Web/API/File) object
and streamUploadFunction being a function that accepts UInt8Arrays and processes them for download (what we're missing), for now what i'm doing is piping the chunks to a new File object (basically rebuilding the file in memory
````javascript
fstream.pipe(cryptor).pipe(new FileWriteStream).on('file',callbackFunction)
```

**Comment by dr-dimitru** [comment]
Sorry for __huge__ delay, this is implemented in [v1.5.1](https://github.com/VeliovGroup/Meteor-Files/releases/tag/1.5.1). Please, (see piping docs)[https://github.com/VeliovGroup/Meteor-Files/wiki/Insert-(Upload)#piping]

**Comment by dr-dimitru** [comment]
As @ghost is removed user, I close this issue
